src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -23140,6 +23140,154 @@ static void M(C c, string? p1, string? p2)
+                // (11,13): warning CS8602: Dereference of a possibly null reference.
+                //         _ = M1(null).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "M1(null)").WithLocation(11, 13));
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/38801")]
+        [WorkItem(37903, "https://github.com/dotnet/roslyn/issues/37903")]
+        public void NotNullIfNotNull_OptionalParameter_LocalFunction()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    void M1()
+    {
+        [return: NotNullIfNotNull(""p"")]
+        string? local1(string? p = ""hello"") => p;
+
+        _ = local1().ToString();
+        _ = local1(null).ToString(); // 1
+        _ = local1(""world"").ToString();
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { NotNullIfNotNullAttributeDefinition, source });
+            comp.VerifyDiagnostics(
+                // (11,13): warning CS8602: Dereference of a possibly null reference.
+                //         _ = local1(null).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "local1(null)").WithLocation(11, 13));
+        }
+
+        [Fact]
+        [WorkItem(37903, "https://github.com/dotnet/roslyn/issues/37903")]
+        public void NotNullIfNotNull_LabeledArguments()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p1"")]
+    string? M1(string? p1, string? p2) => p1;
+
+    void M2()
+    {
+        _ = M1(""hello"", null).ToString();
+        _ = M1(p1: ""hello"", p2: null).ToString();
+        _ = M1(p2: null, p1: ""hello"").ToString();
+        _ = M1(p2: ""world"", p1: null).ToString(); // 1
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { NotNullIfNotNullAttributeDefinition, source });
+            comp.VerifyDiagnostics(
+                // (13,13): warning CS8602: Dereference of a possibly null reference.
+                //         _ = M1(p2: "world", p1: null).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, @"M1(p2: ""world"", p1: null)").WithLocation(13, 13));
+        }
+
+        [Fact]
+        [WorkItem(37903, "https://github.com/dotnet/roslyn/issues/37903")]
+        public void NotNullIfNotNull_LabeledArguments_OptionalParameters()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p1"")]
+    string? M1(string? p1 = ""a"", string? p2 = ""b"") => p1;
+
+    void M2()
+    {
+        _ = M1(p2: null).ToString();
+        _ = M1(p1: null).ToString(); // 1
+        _ = M1(p1: ""hello"", p2: null).ToString();
+        _ = M1(p2: ""hello"", p1: null).ToString(); // 2
+        _ = M1(p1: null, p2: ""hello"").ToString(); // 3
+        _ = M1(p2: null, p1: ""hello"").ToString();
+    }
+}
```

Comment:
Are we testing cases where the optional value is `null`?
```C#
using System.Diagnostics.CodeAnalysis;
public class C
{
    [return: NotNullIfNotNull("p1")]
    string? M1(string? p1 = null, string? p2 = null) => p1;

    void M2()
    {
        _ = M1(p2: "a").ToString(); // 1
        _ = M1(p1: "b").ToString();
        _ = M1(p1: "a", p2: null).ToString();
        _ = M1(p2: "a", p1: null).ToString(); // 2
        _ = M1(p1: null, p2: "b").ToString(); // 3
        _ = M1(p2: null, p1: "b").ToString();
    }
}
```

Reply:
Looks like no, will add the test